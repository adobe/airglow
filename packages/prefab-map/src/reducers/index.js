/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import { call } from '@airglow/reducers';
import * as T from '../actions';
import toggleModeReducer from './toggle.mode.reducer';
import { buttonZoomReducer, zoomReducer } from './zoom.reducer';
import positionReducer from './position.reducer';

export default call(
  call(buttonZoomReducer).for(T.BUTTON_ZOOM),
  call(zoomReducer).for(T.ZOOM),
  call(toggleModeReducer).for(T.TOGGLE_MODE),
  call(positionReducer).for(T.CHANGE_POSITION)
);
