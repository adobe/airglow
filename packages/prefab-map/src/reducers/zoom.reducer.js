/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import * as R from 'ramda';
import { DEFAULT_ZOOM } from '../constants';
import clampZoom from '../util/clamp.map.zoom';

const constructPath = ['construct', 'defaultZoom'];
const zoomPath = ['store', 'zoom'];

const getZoom = state =>
  R.defaultTo(
    R.path(constructPath, state) || DEFAULT_ZOOM,
    R.path(zoomPath, state)
  );
const calcZoom = (current, size) => clampZoom(current + size);

export const buttonZoomReducer = (state, { payload }) =>
  R.assocPath(
    zoomPath,
    calcZoom(getZoom(state), payload.size),
    state
  );

export const zoomReducer = (state, { payload }) =>
  R.assocPath(
    zoomPath,
    payload.zoom,
    state
  );
