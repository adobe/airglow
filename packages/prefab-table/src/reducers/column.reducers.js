/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import * as R from 'ramda';

import columnPrefab from '../column.prefab';

const addedColumnsPath = ['store', 'addedColumns'];

const hiddenColumnsPath = ['store', 'hiddenColumns'];

const addedColumnsLens = R.lensPath(addedColumnsPath);

const hiddenColumnsLens = R.lensPath(hiddenColumnsPath);

export const addColumnReducer = (state, { payload }) => {
  const { column, config } = payload;
  return R.over(addedColumnsLens, R.assoc(column, columnPrefab({ ...config })), state);
};

export const removeColumnReducer = (state, { payload }) => {
  const { column } = payload;
  return R.over(addedColumnsLens, R.dissoc(column), state);
};

export const showColumnReducer = (state, { payload }) => {
  const { column } = payload;
  return R.over(hiddenColumnsLens, R.without([column]), state);
};

export const hideColumnReducer = (state, { payload }) => {
  const { column } = payload;
  const hiddenColumns = R.pipe(
    R.path(hiddenColumnsPath),
    R.append(column),
    R.uniq
  )(state);

  return R.set(hiddenColumnsLens, hiddenColumns, state);
};
