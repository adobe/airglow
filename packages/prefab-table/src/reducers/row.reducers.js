/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import * as R from 'ramda';

const allowsSelectionPath = ['construct', 'allowsSelection'];

const allowsMultipleSelectionPath = ['construct', 'allowsMultipleSelection'];

const selectedRowsPath = ['store', 'selectedRows'];

const selectedRowsLens = R.lensPath(selectedRowsPath);

export const deselectRowsReducer = (state, { payload }) => {
  const { rows } = payload;
  const normalized = !rows
    ? R.view(selectedRowsLens, state)
    : Array.isArray(rows) ? rows : [rows];

  return R.over(selectedRowsLens, R.without(normalized), state);
};

export const toggleRowReducer = (state, { payload }) => {
  if (!R.path(allowsSelectionPath, state)) return state;

  const { row } = payload;

  if (R.indexOf(row, R.pathOr([], selectedRowsPath, state)) >= 0) {
    return R.over(selectedRowsLens, R.without([row]), state);
  }

  return R.path(allowsMultipleSelectionPath, state)
    ? R.over(selectedRowsLens, R.append(row), state)
    : R.set(selectedRowsLens, [row], state);
};
