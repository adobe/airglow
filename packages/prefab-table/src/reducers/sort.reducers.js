/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import * as R from 'ramda';
import { hasColumn } from '../selectors/column.selectors';

const sortPath = ['store', 'sort'];
const defaultSortPath = ['construct', 'defaultSort'];

const getSortState = (state) => {
  const sort = R.path(sortPath, state);
  return sort || R.path(defaultSortPath, state);
}

const sortReducer = (state, { payload }) => {
  const sortState = getSortState(state);
  const { column } = payload;
  let newState = {};
  if (hasColumn(state, column)) {
    if (R.path(['column'], sortState) === column) {
      newState = {
        order: R.negate(R.pathOr(-1, ['order'], sortState)),
        column
      };
    } else {
      newState = {
        order: 1,
        column
      };
    }
  }
  return R.assocPath(
    sortPath,
    R.merge(sortState, newState),
    state
  );
};

export { sortReducer };
