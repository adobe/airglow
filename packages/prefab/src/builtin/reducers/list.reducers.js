/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import * as R from 'ramda';

const valuePath = ['store', 'value'];

const selectValue = state =>
  R.defaultTo(
    R.path(['construct', 'defaultValue'], state),
    R.path(valuePath, state)
  );


const listReducer = (state, { payload }) => R.assocPath(
  valuePath,
  payload.value,
  state
);

const addToListReducer = (state, { payload }) => R.assocPath(
  valuePath,
  R.append(payload.value, selectValue(state)),
  state
);

const removeFromListReducer = (state, { payload }) => R.assocPath(
  valuePath,
  R.without([payload.value], selectValue(state)),
  state
);

const removeAtIndexReducer = (state, { payload }) => R.assocPath(
  valuePath,
  R.remove(payload.value, payload.size || 1, selectValue(state)),
  state
);

export { listReducer, addToListReducer, removeFromListReducer, removeAtIndexReducer };
