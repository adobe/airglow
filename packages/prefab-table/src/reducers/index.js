/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import { call } from '@airglow/reducers';
import * as T from '../actions';
import { addColumnReducer, hideColumnReducer, removeColumnReducer, showColumnReducer } from './column.reducers';
import { deselectRowsReducer, toggleRowReducer } from './row.reducers';
import { sortReducer } from './sort.reducers';

export default call(
  call(addColumnReducer).for(T.ADD_COLUMN),
  call(deselectRowsReducer).for(T.DESELECT_ROWS),
  call(hideColumnReducer).for(T.HIDE_COLUMN),
  call(removeColumnReducer).for(T.REMOVE_COLUMN),
  call(showColumnReducer).for(T.SHOW_COLUMN),
  call(sortReducer).for(T.DO_SORT),
  call(toggleRowReducer).for(T.TOGGLE_ROW)
);
