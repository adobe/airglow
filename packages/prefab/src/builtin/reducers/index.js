/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/


import { call } from '@airglow/reducers';
import * as T from '../actions';

import {
  turnOnReducer, turnOffReducer, toggleReducer
} from './toggle.reducers';

import {
  valueReducer, resetReducer
} from './value.reducers';

import {
  listReducer, addToListReducer, removeFromListReducer, removeAtIndexReducer
} from './list.reducers';


export const reduceToggle = call(
  call(turnOnReducer).for(T.TOGGLE_ON),
  call(turnOffReducer).for(T.TOGGLE_OFF),
  call(toggleReducer).for(T.TOGGLE_SWITCH)
);

export const reduceValue = call(
  call(valueReducer).for(T.VALUE),
  call(resetReducer).for(T.VALUE_RESET)
);

export const reduceList = call(
  call(listReducer).for(T.LIST_SET),
  call(addToListReducer).for(T.LIST_ADD),
  call(removeFromListReducer).for(T.LIST_REMOVE),
  call(removeAtIndexReducer).for(T.LIST_REMOVE_AT)
);
