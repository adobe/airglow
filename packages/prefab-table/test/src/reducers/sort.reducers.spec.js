/*
 * ************************************************************************
 * ADOBE CONFIDENTIAL
 * ___________________
 *
 *   Copyright 2017 Adobe Systems Incorporated
 *   All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Adobe Systems Incorporated and its suppliers,
 * if any.  The intellectual and technical concepts contained
 * herein are proprietary to Adobe Systems Incorporated and its
 * suppliers and are protected by all applicable intellectual property
 * laws, including trade secret and copyright laws.
 * Dissemination of this information or reproduction of this material
 * is strictly forbidden unless prior written permission is obtained
 * from Adobe Systems Incorporated.
 * ************************************************************************
 */

import { sortReducer } from '../../../src/reducers/sort.reducers';

const payload = column => ({ payload: {
  construct: 'fruit',
  column
} });

const state = (order, column) => ({
  construct: {
    columns: {
      name: {},
      color: {}
    }
  },
  store: {
    sort: {
      order,
      column
    }
  }
});

const initState = {
  construct: {
    columns: {
      name: {},
      color: {}
    }
  },
  store: {
  }
};

const defaultState = {
  construct: {
    columns: {
      name: {},
      color: {}
    },
    defaultSort: {
      column: 'name',
      order: 1
    }
  },
  store: {
  }
};


describe('sortReducer', () => {
  it('should ignore', function () {
    expect(sortReducer(initState, payload('type')))
      .toMatchSnapshot();
  });
  it('should set column and order to 1 from empty state', function () {
    expect(sortReducer(initState, payload('name')))
      .toMatchSnapshot();
  });
  it('should use the default values', function () {
    expect(sortReducer(defaultState, payload('name')))
      .toMatchSnapshot();
  });
  it('should toggle sort order for same column', function () {
    expect(sortReducer(state(1, 'name'), payload('name')))
      .toMatchSnapshot();
  });
  it('should sort new column', function () {
    expect(sortReducer(state(1, 'name'), payload('color')))
      .toMatchSnapshot();
  });
});
