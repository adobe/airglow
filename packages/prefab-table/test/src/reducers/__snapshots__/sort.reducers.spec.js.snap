// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sortReducer should ignore 1`] = `
Object {
  "construct": Object {
    "columns": Object {
      "color": Object {},
      "name": Object {},
    },
  },
  "store": Object {
    "sort": Object {},
  },
}
`;

exports[`sortReducer should set column and order to 1 from empty state 1`] = `
Object {
  "construct": Object {
    "columns": Object {
      "color": Object {},
      "name": Object {},
    },
  },
  "store": Object {
    "sort": Object {
      "column": "name",
      "order": 1,
    },
  },
}
`;

exports[`sortReducer should sort new column 1`] = `
Object {
  "construct": Object {
    "columns": Object {
      "color": Object {},
      "name": Object {},
    },
  },
  "store": Object {
    "sort": Object {
      "column": "color",
      "order": 1,
    },
  },
}
`;

exports[`sortReducer should toggle sort order for same column 1`] = `
Object {
  "construct": Object {
    "columns": Object {
      "color": Object {},
      "name": Object {},
    },
  },
  "store": Object {
    "sort": Object {
      "column": "name",
      "order": -1,
    },
  },
}
`;
